kclvm/compiler_base/diagnostic/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+///  |     ^ error here!
+/// error: aborting due to previous error.
+/// For more information about this error.
+/// 
+/// It consists of 4 'Sentence'
+/// 
+/// Sentence 1: error[E0999]: oh no! this is an error!
+/// 
+/// Sentence 2: 
+/// --> mycode.rs:3:5
+///  |
+///3 |     a:int
+///  |     ^ error here!
+/// 
+/// Sentence 3: error: aborting due to previous error.
+/// 
+/// Sentence 4: For more information about this error.
+pub struct Diagnostic {
+    pub messages: Vec<Sentence>,
+}
+
+impl Diagnostic {
+    pub fn new() -> Self {
+        Diagnostic { messages: vec![] }
+    }
+
+    pub fn add_sentence(&mut self, sentence: Sentence) {
+        self.messages.push(sentence);
+    }
+}
+
+pub trait DiagnosticBuilder {
+    fn into_diagnostic(self) -> Diagnostic;
+}
+
+/// 'Sentence' consists of 'Pendant' and sentence content.
+/// 'Pendant' is optional.
+/// 
+/// e.g. 
+/// Sentence 1: error[E0999]: oh no! this is an error!
+///     Pendant: error[E0999]:
+///     Sentence content: oh no! this is an error!
+/// 
+/// Sentence 2: 
+/// --> mycode.rs:3:5
+///  |
+///3 |     a:int
+///  |     ^ error here!
+///     Pendant: 
+///     --> mycode.rs:3:5
+///       |
+///     3 |     a:int
+///       |     ^ 
+///     Sentence content: error here!
+/// 
+/// Sentence 3: error: aborting due to previous error.
+///     Pendant: error
+///     Sentence content: aborting due to previous error.
+/// 
+/// Sentence 4: For more information about this error.
+///     Pendant: no pendant
+///     Sentence content: For more information about this error.
+pub struct Sentence {
+    pendant: Box<dyn Pendant>,
+    sentence: Message,
+}
+
+pub trait Pendant {
+    fn format(&self, shader: Rc<dyn Shader>, sb: &mut StyledBuffer);
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum Message {
+    // 'Sentence' text message
+    Str(String),
+    // 'Sentence' text template id
+    // TODO(zong-zhe): Some text messages are too long and require template.
+    FluentId(String),
+}
+
+impl Sentence {
+    pub fn new_sentence_str(pendant: Box<dyn Pendant>, sentence: Message) -> Self {
+        Self { pendant, sentence }
+    }
+
+    pub fn new_nopendant_sentence(sentence: Message) -> Self {
+        Self {
+            pendant: Box::new(NoPendant::new()),
+            sentence,
+        }
+    }
+
+    pub fn format(&self, shader: Rc<dyn Shader>, sb: &mut StyledBuffer) {
+        let sentence_style = shader.normal_msg_style();
+        self.pendant.format(shader, sb);
+        match &self.sentence {
+            Message::Str(s) => sb.appendl(&s, sentence_style),
+            Message::FluentId(s) => sb.appendl(&s, sentence_style.clone()),
+        }
+    }
+}
+
+/// Position describes an arbitrary source position including the filename,
+/// line, and column location.
+///
+/// A Position is valid if the line number is > 0.
+/// The line and column are both 1 based.
+#[derive(PartialEq, Clone, Eq, Hash, Debug, Default)]
+pub struct Position {
```

Comment:
Use `Span` and `SourceMap` instead of Position to calculate the error `Postion`.